semantics/src/main/kotlin/com/strumenta/kolasu/semantics/scope/description/ScopeDescription.kt

```diff
@@ -4,133 +4,148 @@ import com.strumenta.kolasu.model.Node
- **/
-fun scope(
-    ignoreCase: Boolean = false,
-    init: ScopeDescriptionApi.() -> Unit
-): ScopeDescription = ScopeDescription(ignoreCase).apply(init)
-
-/**
- * A scope description containing all relevant information
- * concerning the visible elements at a given point in space or time.
- * - `ignoreCase`: whether the names associated to the symbols
- * in this scope should be handled as case-sensitive or not;
- * - `nodes`: the local symbols contained in this scope and
- * their associated name;
- * - `identifiers`: the external symbols contained in this scope
- * and their associated name;
- * - `parent`: the scope description *preceding* the current one - used
- * as delegate if no symbol can be found for a given name in the current scope;
+ * Description of the visible names at a given point in time-space.
  *
- * Each scope description can be configured through the methods defined in
- * the `ScopeDescriptionApi`. References can be resolved invoking `resolve(reference)`.
- *
- * Given a reference, the scope provider will follow the following plan:
- * - try to retrieve a local symbol associated with the given reference name;
- * - if not found, try to retrieve a global symbol associated with the given reference name;
- * - if not found, delegate the resolution to its parent (if any)
+ * @param ignoreCase flag indicating that name resolution should be case-insensitive (if true)
+ * @param parent the parent [ScopeDescription] containing this scope description
  *
+ * @author Lorenzo Addazi <lorenzo.addazi@strumenta.com>
  **/
 class ScopeDescription(
-    private val ignoreCase: Boolean = false
-) : ScopeDescriptionApi {
-    private var parent: ScopeDescription? = null
-    private val namesToExternalSymbolIdentifiers: MutableMap<String, String> = mutableMapOf()
-    private val namesToLocalSymbolNodes: MutableMap<String, PossiblyNamed> = mutableMapOf()
+    var ignoreCase: Boolean = false,
+    var parent: ScopeDescription? = null
+) {
 
     /**
-     * Resolves the given reference in the current scope (or its parents).
+     * Map associating symbol names with case-insensitive keys - used to keep
+     * case information about symbols (might change in keys when [ignoreCase] is true).
      **/
-    fun resolve(reference: ReferenceByName<out PossiblyNamed>) {
-        val name = reference.name.asKey()
-        val node by lazy { this.namesToLocalSymbolNodes[name] }
-        val identifier by lazy { this.namesToExternalSymbolIdentifiers[name] }
-        when {
-            node != null -> {
-                @Suppress("UNCHECKED_CAST")
-                (reference as ReferenceByName<PossiblyNamed>).referred = node
-            }
-            identifier != null -> {
-                reference.identifier = identifier
-            }
-            else -> {
-                this.parent?.resolve(reference)
-            }
-        }
-    }
+    private val namesToSymbolKeys: MutableMap<String, String> = mutableMapOf()
```

Comment:
Can't I have more than one symbol with the same name in the scope? For example, overloaded methods

Reply:
Yes, multiple symbols for the same symbol key are possible - both in the local and external symbols map. This map is only keeping track of the original name and the corresponding key. For example, if I have a class 'Person' and a variable 'person' within a case-insensitive scope description, there will be two entries in this map: 'Person' -> 'person' and 'person' -> 'person'. The actual symbols will both be associated with the key 'person'. This allows us to keep track of the original names, which are used in features like code completion. Otherwise, we would lose this information when converting names to keys, e.g. Person would become person and be suggested in the wrong way.

Reply:
Considering the overloaded methods support: it would require a specific symbol describing methods. I think this is a language-specific feature. In the scope description, we can have multiple symbols associated to a given name. However, we can only have one symbol per type in this list. Hence, the multiple overloaded implementations of a method should be represented inside the function symbol. Am I wrong?

Reply:
> Hence, the multiple overloaded implementations of a method should be represented inside the function symbol. Am I wrong?

I am not sure I understand this part

Reply:
From my point of view symbol resolution should figure out, among other things, which method get called. And that means also distinguishing between overloaded variants of methods with the same name. So after symbol resolution is performed I would expect the reference in a method call to point to the right method declaration. In this way other systems can just consume that information without having to know anything on the method resolution rules. For example, some editor could just use that information to implement a "go to declaration" functionality. Symbol resolution should fine the information necessary to perform the resolution by looking at the symbol definition.

> Otherwise, how should handle cases where multiple symbols of the same type and name are contained in a given scope? 

I was thinking to have a symbol for each overloaded variant, describing the signature of that method and symbol resolution would consider the information to pick the right symbol. If the actual symbol resolution is delegated to FunctionSymbol, then the user needs to know about it and calling the appropriate method. 

> Otherwise, how should handle cases where multiple symbols of the same type and name are contained in a given scope? 

Considering not only the name but the whole symbol description. Otherwise why do we need to have any information but the name in the Symbol Description? The information is there only to be consumed during symbol resolution, or at least this was my understanding

Reply:
At the moment, the symbol resolver traverses the AST and aggregates all references. For each reference, it computes the corresponding scope and tries to resolve the reference using this. The resolution is generic and uses:

- The name of the reference by name instance;
- The type of the reference by name definition;

It absolutely makes sense to consider the signature in cases of overloaded functions, for example. I am not sure, however, how this could be implemented in a generic way. It looks to me like we might need some mechanism to specify the matching criteria. For example, in the case of a variable, it might be the name. In case of functions, it could integrate parameter type information as well.

The actual sequence of symbol resolution steps might change from node to node. For example, when resolving symbols for a function call expression, we might want to do it on its argument expressions first. I feel we could introduce support to specify resolution steps, for example:

```kotlin
resolve(InvocationExpr::class) { (invocationExpr) ->
   // resolve symbols in arguments first
   invocationExpr.arguments.forEach { argument -> resolve(argument) }
   // compute the scope for the invocationExpr::function property
   val scope = scopeProvider.scopeFor(invocationExpr.function)
   // use the scope to obtain a list of candidates
   // select the most appropriate candidate according to some criteria
   invocationExpr.function.identifier = /* the identifier of the selected candidate */
   // we could also differentiate between externalCandidates and localCandidates
   invocationExpr.function.referred = /* the node result from asking local candidates */
}
```

The workflow would consist in:
- Defining a Symbol provider to specify which nodes can be referenced from the outside;
- Defining a Scope provider to specify what is visible from a given Node;
- Defining a Type provider to specify the type for a given Node (considers references as resolved);
- Defining a Symbol resolver to specify how to resolve symbols;

In the definition of a symbol resolution rule, we might use a type provider to retrieve the type of a given node - with the assumption that all required symbols have been resolved if possible. A scope provider to return back the scope description from a given node. The scope provider can use the type provider, but the type provider cannot use the scope provider. Symbol resolution rules coordinate and make sure that symbols are resolved before being used. Does this make sense?